### ISSUE TITLE
New Rule: Prevent hidden garbage collection issues

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#18823]
### Rule details

The rule should flag patterns that may result in variables not being garbage collected

### Related ECMAScript feature

(All)

### What type of rule is this?

Warns about a potential problem

### Example code

```js
function demo() {

  // this variable isn't GCed because it's referenced in a closure and there's another function in scope
  const bigArrayBuffer = new ArrayBuffer(100_000_000);
  const id = setTimeout(() => {
    console.log(bigArrayBuffer.byteLength);
  }, 1000);

  return () => clearTimeout(id);
}

globalThis.cancelDemo = demo();
```

### Why should this rule be in the core instead of a plugin?

This pattern is particularly difficult to notice and can have significant negative performance impact, especially if you're dealing with a large amount of data.

Writeup here:
https://jakearchibald.com/2024/garbage-collection-and-closures/

According to the post, this looks unlikely to be solved in engines due to performance impact.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by nzakas** [comment]
@eslint/eslint-team still looking for thoughts on this.

**Comment by JoshuaKGoldberg** [comment]
🤔 This seems like a very niche, difficult-to-detect case to me. At least the following two points of difficulty:

* It'd require detecting code constructs that call to multiple APIs (buffers, timing) in a specific way
* Misdirection in code, such as utility functions to do ☝, would work around the rule quickly

It's also not very common to use `globalThis` variables. So even if this was a rule that could be written in a relatively straightforward manner, it wouldn't make a positive impact on many apps.

The linked post references https://www.schiener.io/2024-03-03/react-closures showing how to trigger this in React. But there'd be no way for ESLint to know that a particular frameworks's functions (e.g. `useCallback`) preserve references in a particular way.

My vote would be to not implement this in ESLint core, with an encouraging note that it'd be great to try out in a userland plugin.

**Comment by nzakas** [comment]
@JoshuaKGoldberg I think you may be overthinking this pattern. It's not about `globalThis` at all, it's about closures referencing variables from an upper scope inside of a function that disappears. Those first two parts (detecting a closure and detecting a disappearing function) are relatively straightforward to detect. The more difficult part would be to determine if the function was freed from memory later.

Still, it seems worth taking a look because this pattern is difficult to see on its own.

@eslint/eslint-tsc still looking for feedback here.

**Comment by fasttime** [comment]
I am able to verify the problem described in this issue in Node.js v22 by forcing garbage collection with [`gc()`](https://nodejs.org/api/cli.html#--expose-gc) and measuring memory usage with [`process.memoryUsage()`](https://nodejs.org/api/process.html#processmemoryusage):

```js
// test.mjs -- run in Node.js v22.3.0 or later with `node --expose-gc test.mjs`
function demo() {

    // this variable isn't GCed because it's referenced in a closure and there's another function in scope
    const bigArrayBuffer = new ArrayBuffer(100_000_000);
    const id = setTimeout(() => {
      console.log(bigArrayBuffer.byteLength);
    }, 1000);
  
    return () => clearTimeout(id);
}
  
globalThis.cancelDemo = demo();

import timers from "node:timers/promises";
await timers.setTimeout(2000);
gc(); // run garbage collection
console.log(process.memoryUsage());
```

I'm getting this output:

```console
100000000
{
  rss: 40632320,
  heapTotal: 6422528,
  heapUsed: 4517216,
  external: 100433823,
  arrayBuffers: 100016762
}
```

The line with `arrayBuffers: 100016762` indicates that the large ArrayBuffer object was not garbage-collected.

**Comment by fasttime** [comment]
The example provided in the blog works because `setTimeout` associates the callback to a generated id using a host internal map, and this mapping prevents the callback from being garbage-collected after the timeout expires.

If `setTimeout` is replaced by a different call like `setImmediate` in Node.js, the ArrayBuffer will be no longer retained, as I was able to verify using `process.memoryUsage()` (see my previous comment).

This means that a new rule would need knowledge of what a function is doing with a callback in order to determine if there's an issue with garbage collection. For well-known functions like HTML's `setTimeout` or React's `useCallback` hook this should be possible in most cases, but there will be both false positives and false negatives.

**Comment by nzakas** [comment]
All right, it doesn't seem like there's much interest in pursuing this, so closing.
